### PR TITLE
Timeout lpadmin command issue#41

### LINKED ISSUE #41: Login-freezes: Set Timeout by lpadmin-command
Problem: Login freezes if CUPS is not a local CUPS and user has AD-printers
use timeout with e.g. 5 seconds
File: printers.py
```
def _installPrinter(username, networkPath, name):
    logging.debug("_installPrinter Printer {0} on {1}".format(name, networkPath))
    installCommand = ["timeout", "5", "lpadmin", "-p", name, "-E", "-v", networkPath, "-m", "everywhere", "-u", f"allow:{username}"]
    logging.debug("* running '{}'".format(" ".join(installCommand)))
    p = subprocess.Popen(installCommand, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    output, error = p.communicate()
    if p.returncode  == 0:
        logging.debug("* Success _installPrinter!")
        return True
    elif p.returncode == 124:
        logging.debug("* Error Timeout lpadmin")
    else:
        logging.debug(f"* Error installing printer {name} on {networkPath}!\n")
        logging.debug(f"* stderr: {error}")
    return False
```

### PR DESCRIPTION
**Pull request by martin-res**
Timeout ist auf 3 Sekunden gesetzt, kann ggf. geändert werden
bei timeout erhält man returncode 124
timeout-check für "lpstat" m.E. nicht notwendig
closes #41

### CODE REVIEW

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 1]
```diff
@@ -73,16 +73,20 @@ def translateSambaToIpp(networkPath):
 # --------------------
 
 def _installPrinter(username, networkPath, name):
-    logging.debug("Installing Printer {0} on {1}".format(name, networkPath))
-    installCommand = ["lpadmin", "-p", name, "-E", "-v", networkPath, "-m", "everywhere", "-u", f"allow:{username}"]
+    logging.info("_installPrinter {0} on {1}".format(name, networkPath))
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:76]: Why did you change that?

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 2]
```diff
@@ -73,16 +73,20 @@ def translateSambaToIpp(networkPath):
...
+    installCommand = ["timeout", "3", "lpadmin", "-p", name, "-E", "-v", networkPath, "-m", "everywhere", "-u", f"allow:{username}"]
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:77]: I think, 3 seconds may be too low in some cases of slow networks. Please change this to 15 Seconds.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 3]
```diff
@@ -73,16 +73,20 @@ def translateSambaToIpp(networkPath):
...
     logging.debug("* running '{}'".format(" ".join(installCommand)))
-
-    if not subprocess.call(installCommand, stdout=subprocess.PIPE, stderr=subprocess.PIPE) == 0:
-        logging.fatal(f"* Error installing printer {name} on {networkPath}!\n")
-        return False
-
-    logging.debug("* Success!")
-    return True
+    p = subprocess.Popen(installCommand, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:79]: Is it necessary to use popen here? Is there a case of some important output?

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 4]
```diff
@@ -73,16 +73,20 @@ def translateSambaToIpp(networkPath):
...
+    output, error = p.communicate()
+    if p.returncode  == 0:
+        logging.debug("* Success _installPrinter!")
+        return True
+    elif p.returncode == 124:
+        logging.debug("* Error Timeout lpadmin")
+    else:
+        logging.debug(f"* Error installing printer {name} on {networkPath}!\n")
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:87]: Please change this to logging.fatal. Otherwise no error popup will be shown.
```suggestion
        logging.fatal(f"* Error installing printer {name} on {networkPath}!\n")
```

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 5]
```diff
@@ -73,16 +73,20 @@ def translateSambaToIpp(networkPath):
...
-
-    logging.debug("* Success!")
-    return True
+    p = subprocess.Popen(installCommand, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    output, error = p.communicate()
+    if p.returncode  == 0:
+        logging.debug("* Success _installPrinter!")
+        return True
+    elif p.returncode == 124:
+        logging.debug("* Error Timeout lpadmin")
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:85]:
```suggestion
        logging.error(f"* Timeout error while installing printer {name} on {networkPath}")
```

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 6]
```diff
@@ -113,12 +117,17 @@ def _getInstalledPrintersOfUser(username):
...
-    logging.info("* Uninstalling Printer {}".format(name))
-    uninstallCommand = ["lpadmin", "-x", name]
-    #logging.debug("* running '{}'".format(installCommand))
-
-    if not subprocess.call(uninstallCommand) == 0:
-        logging.error("* Error uninstalling printer!")
-        return False
-
-    return True
+    logging.info("* _uninstallPrinter {}".format(name))
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:120]: Is all of this necessary over here? I have never seen a freeze during uninstalling a printer.

**Review by dorianim — CHANGES_REQUESTED** [review]

**Comment by dorianim** [comment]
Also, please add `coreutils` to the package dependencoes to make sure, the timeout command is available 🙂

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 5 above)
Comment (martin-res) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:85]: OK

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 2 above)
Comment (martin-res) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:77]: 15 Sekunden halte ich für zu lang, da dann solange der login-Prozess blockiert wird, ggf, mehrfach bei mehreren Druckern. 3 Sekunden genügte bei meiner langsamen virtuellen Umgebung, ändern auf 5 Sekunden?

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 1 above)
Comment (martin-res) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:76]: wegen Konsistenz zu anderen Funktionsaufrufen, die auch .info verwenden

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 3 above)
Comment (martin-res) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:79]: output: stderr for debugging

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 6 above)
Comment (martin-res) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:120]: Bin mir nicht sicher, hatte beim Testen auch den Fall, dass beim Deinstallieren ein Passwort abgefragt wurde.

Comment (dorianim) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:120]: Could you reproduce that again?
I don't think this can happen, since the installation would fail in that case, so the printer would not be uninstalled.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 2 above)
Comment (dorianim) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:77]: I think we should at least take 10 seconds. This is just an edge case and should never happen in a properly configured environment.
But we could make it configurable once #32 is done.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 1 above)
Comment (dorianim) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:76]: Could you please provide an example? I cannot find an occurrence of something like this anywhere else.
Furthermore, `_installPrinter` is an internal function (indicated by the `_`) and should therefore not be mentioned in logs, they are meant for users, not for developers.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 3 above)
Comment (dorianim) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:79]: Yes, but what kind of useful information does lpadmin output? I think it is unnecessarily complicated to use popen here.

Comment (dorianim) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:79]: The output is recorded in the syslog anyway. We don't need to capture it over here.

Comment (martin-res) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:79]: Für die Funktionalität ist es gleichwertig, es erleicherte mir das Debuggen und ist mehrfach getestet. Auch für zukünftiges Debuggen würde ich es beibehalten.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 6 above)
Comment (martin-res) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py]: Bei einem Server-Cups werden bei uninstall alle dort aufgelisteten Drucker abgearbeitet, auch wenn sie vorher nicht installiert wurden, insofern sicherheitshalber notwendig.

**Comment by martin-res** [comment]
Habe auch logging-Messages noch angepasst - muss ich nun einen neuen PR starten?

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 3 above)
Comment (dorianim) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:79]: We use popen nowhere else. The reason being (as I already said), that all the logs are present in the syslog anyway and can be seen in linuxmuster-linuxclient7 print-logs. If we change it here, we would also have to change it in all of these places:
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/19e41e02aae7409baec47a1a5f8a9caaae414ea8/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/shares.py#L256
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/19e41e02aae7409baec47a1a5f8a9caaae414ea8/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/shares.py#L243
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/19e41e02aae7409baec47a1a5f8a9caaae414ea8/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/shares.py#L203
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/19e41e02aae7409baec47a1a5f8a9caaae414ea8/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/shares.py#L177
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/3eb7d17ec0232d51ac194a30e45399f0bf908379/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/realm.py#L17
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/3eb7d17ec0232d51ac194a30e45399f0bf908379/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/realm.py#L35
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L76
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L80
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L84
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L88
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L102
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L103
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L138
- https://github.com/linuxmuster/linuxmuster-linuxclient7/blob/2ece10c6957002090e227540e1badda94a9c5808/usr/lib/python3/dist-packages/linuxmusterLinuxclient7/imageHelper.py#L209

And probably even more. This could be done but would require separate PR. So please, use `subprocess.call` just as before.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 7]
```diff
@@ -73,16 +73,20 @@ def translateSambaToIpp(networkPath):
...
     logging.debug("* running '{}'".format(" ".join(installCommand)))
-
-    if not subprocess.call(installCommand, stdout=subprocess.PIPE, stderr=subprocess.PIPE) == 0:
+    p = subprocess.Popen(installCommand, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    output, error = p.communicate()
+    if p.returncode  == 0:
+        logging.debug("* Success Install Printer!")
+        return True
+    elif p.returncode == 124:
+        logging.debug("* Error timeout lpadmin")
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:85]:
```suggestion
        logging.fatal(f"* Timeout error while installing printer {name} on {networkPath}")
```
The log should be fatal and contain some more information.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 8]
```diff
@@ -73,16 +73,20 @@ def translateSambaToIpp(networkPath):
...
+    else:
         logging.fatal(f"* Error installing printer {name} on {networkPath}!\n")
-        return False
-
-    logging.debug("* Success!")
-    return True
+        # logging.debug(f"* stderr: {error}")
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:88]:
```suggestion
```

Please remove this comment.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 9]
```diff
@@ -113,12 +117,17 @@ def _getInstalledPrintersOfUser(username):
...
-
-    if not subprocess.call(uninstallCommand) == 0:
-        logging.error("* Error uninstalling printer!")
-        return False
-
-    return True
+    logging.info("Uninstall Printer {}".format(name))
+    uninstallCommand = ["timeout", "10", "lpadmin", "-x", name]
+    logging.debug("* running '{}'".format(" ".join(uninstallCommand)))
+    p = subprocess.Popen(uninstallCommand, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:123]: Same as above. Please change to `sobprocess.call`, like it was before.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 10]
```diff
@@ -113,12 +117,17 @@ def _getInstalledPrintersOfUser(username):
...
+    output, error = p.communicate()
+    if p.returncode  == 0:
+        logging.debug("* Success Uninstall Printer!")
+        return True
+    elif p.returncode == 124:
+        logging.debug("* Error timeout lpadmin")
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:129]: Same as above
```suggestion
        logging.fatal(f"* Timeout error while installing printer {name} on {networkPath}")
```

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 11]
```diff
@@ -113,12 +117,17 @@ def _getInstalledPrintersOfUser(username):
...
+    else:
+        logging.debug(f"* Error Uninstall Printer {name}!")
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:131]:
```suggestion
        logging.fatal(f"* Error Uninstall Printer {name}!")
```
This should be fatal.

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py [hunk 12]
```diff
@@ -113,12 +117,17 @@ def _getInstalledPrintersOfUser(username):
...
+        # logging.debug(f"* stderr: {error}")
```
Comment (dorianim) [inline usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py:132]:
```suggestion
```

Please remove this comment.

**Review by dorianim — CHANGES_REQUESTED** [review]

File: usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py (hunk 6 above)
Comment (dorianim) [reply usr/lib/python3/dist-packages/linuxmusterLinuxclient7/printers.py]: Ok, whatever. Let's leave it in, should not hurt anybody.

**Comment by dorianim** [comment]
You don't need a new PR. I will merge this as soon as all conversations are resolved, and I'm happy with the code.
By the way: You can directly commit proposed changes:
image

**Review by dorianim — APPROVED** [review]
Thank you.
